data-prepper-plugins/flatten-json-processor/src/main/java/org/opensearch/dataprepper/plugins/processor/flattenjson/FlattenJsonProcessor.java

```diff
@@ -0,0 +1,138 @@
+/*
+ * Copyright OpenSearch Contributors
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+package org.opensearch.dataprepper.plugins.processor.flattenjson;
+
+import com.github.wnameless.json.flattener.JsonFlattener;
+import org.opensearch.dataprepper.expression.ExpressionEvaluator;
+import org.opensearch.dataprepper.metrics.PluginMetrics;
+import org.opensearch.dataprepper.model.annotations.DataPrepperPlugin;
+import org.opensearch.dataprepper.model.annotations.DataPrepperPluginConstructor;
+import org.opensearch.dataprepper.model.event.Event;
+import org.opensearch.dataprepper.model.processor.AbstractProcessor;
+import org.opensearch.dataprepper.model.processor.Processor;
+import org.opensearch.dataprepper.model.record.Record;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.ArrayList;
+import java.util.Collection;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+
+@DataPrepperPlugin(name = "flatten", pluginType = Processor.class, pluginConfigurationType = FlattenJsonProcessorConfig.class)
+public class FlattenJsonProcessor extends AbstractProcessor<Record<Event>, Record<Event>> {
+    private static final Logger LOG = LoggerFactory.getLogger(FlattenJsonProcessor.class);
+
+    private static final String SEPARATOR = "/";
+    private final FlattenJsonProcessorConfig config;
+    private final ExpressionEvaluator expressionEvaluator;
+
+    @DataPrepperPluginConstructor
+    public FlattenJsonProcessor(final PluginMetrics pluginMetrics, final FlattenJsonProcessorConfig config, final ExpressionEvaluator expressionEvaluator) {
+        super(pluginMetrics);
+        this.config = config;
+        this.expressionEvaluator = expressionEvaluator;
+    }
+
+    @Override
+    public Collection<Record<Event>> doExecute(final Collection<Record<Event>> records) {
+        for (final Record<Event> record : records) {
+            final Event recordEvent = record.getData();
+
+            try {
+                if (config.getFlattenWhen() != null && !expressionEvaluator.evaluateConditional(config.getFlattenWhen(), recordEvent)) {
+                    continue;
+                }
+
+                final String sourceJson = recordEvent.getAsJsonString(config.getSource());
+
+                // adds ignoreReservedCharacters() so that dots in keys are ignored during flattening
+                // e.g., {"a.b": {"c": 1}} will be flattened as expected: {"a.b.c": 1}; otherwise, flattened result will be {"[\"a.b\"]c": 1}
+                Map<String, Object> flattenedJson = new JsonFlattener(sourceJson).ignoreReservedCharacters().flattenAsMap();
+
+                if (config.isRemoveProcessedFields()) {
+                    final Map<String, Object> sourceMap = recordEvent.get(config.getSource(), Map.class);
+                    for (final String keyInSource : sourceMap.keySet()) {
+                        recordEvent.delete(getJsonPointer(config.getSource(), keyInSource));
+                    }
```

Comment:
So we are only worried about supporting flattening one level below the source key that is configured?

Reply:
The flattening will go as deep as needed. The deletion should only need to operate one level below source key to remove processed fields (all fields under source are processed fields) - which is essentially "clearing" the source field, making it an empty map.